File: android/src/main/java/com/bluetoothle/PromiseManager.kt
package com.bluetoothle

import com.bluetoothle.GlobalOptions.Companion.keys.timeoutDuration
import com.bluetoothle.types.Error
import com.bluetoothle.types.PromiseType
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeArray

class PromiseManager(private val pendingPromises: HashMap<PromiseType, MutableList<Pair<Promise, Timeout?>>>) {
  fun resolvePromise(promiseType: PromiseType, payload: WritableMap) {
    val (promise, timeout) = getNextPromise(promiseType) ?: return

    timeout?.cancel()
    promise.resolve(payload)
  }

  fun addPromise(promiseType: PromiseType, promise: Promise?, timeout: Int?) {
    if (promise == null) return

    if (pendingPromises[promiseType] == null) {
      pendingPromises[promiseType] = mutableListOf()
    }

    val promiseTimeout = getTimeout(promiseType, timeout)
    pendingPromises[promiseType]?.add(Pair(promise, promiseTimeout))
  }

  private fun getTimeout (promiseType: PromiseType, duration: Int?): Timeout? {
    if (duration == null || duration <= 0) {
      return null
    }

    val onTimeout = Runnable {
      val response = Arguments.createMap().apply { putString(Strings.error, Error.TIMEOUT.toString()) }
      resolvePromise(promiseType, response)
    }

    val timeout = Timeout()
    timeout.set(onTimeout, duration)
    return timeout
  }

  private fun getNextPromise(promiseType: PromiseType): Pair<Promise, Timeout?>? {
    val list = pendingPromises[promiseType]

    if (list.isNullOrEmpty()) {
      return null
    }

    val nextPromise = list.first()
    list.removeFirst()

    return nextPromise
  }

  fun resolveAllWithError(error: Error) {
    val response = Arguments.createMap().apply {
      putString(Strings.error, error.toString())
    }

    pendingPromises.values.forEach(){ promiseList ->
      promiseList.forEach{item ->
        val (promise, timeout) = item
        timeout?.cancel()
        promise.resolve(response)
      }
    }
  }
}

File: PromiseManager.swift
//
//  Promises.swift
//  react-native-bluetooth-le
//
//  Created by Andrew Tkachuk on 04.01.2024.
//

import Foundation

class PromiseManager: NSObject {
    var pendingPromises = [PromiseType : [(RCTPromiseResolveBlock, Timeout?)]]()
    
    override init() {
    }
    
    func resolvePromise(promiseType: PromiseType, payload: NSDictionary) {
        guard let (resolve, timeout) = getNextPromise(promiseType: promiseType) else {return}
        
        timeout?.cancel()
        resolve(payload)
    }
    
    func addPromise (promiseType: PromiseType, promise: RCTPromiseResolveBlock?, timeout: Int?) {
        if (promise == nil) {
            return
        }
        
        if (pendingPromises[promiseType] == nil) {
            pendingPromises[promiseType] = []
        }
        
        let promiseTimeout = getTimeout(promiseType: promiseType, duration: timeout)
        pendingPromises[promiseType]?.append((promise!, promiseTimeout))
    }
    
    private func getTimeout (promiseType: PromiseType, duration: Int?) -> Timeout? {
        guard let timeoutDuration = duration, timeoutDuration > 0 else {return nil}
        
        func onTimeout () {
            resolvePromise(promiseType: promiseType, payload: [Strings.error: ErrorMessage.TIMEOUT.rawValue] as NSDictionary)
        }
        
        let timeout = Timeout()
        timeout.set(callback: onTimeout, duration: timeoutDuration)
        return timeout
    }
    
    
    private func getNextPromise(promiseType: PromiseType) -> (RCTPromiseResolveBlock, Timeout?)? {
        guard var list = pendingPromises[promiseType], let nextPromise = list.first else {return nil}
        
        list.removeFirst()
        pendingPromises[promiseType] = list
        
        return nextPromise
    }
    
    func resolveAllWithError (error: ErrorMessage) {
        let response = [Strings.error: error.rawValue]
        
        for items in pendingPromises.values {
            for item in items {
                let (promise, timeout) = item
                timeout?.cancel()
                promise(response)
            }
        }
    }
}

enum PromiseType {
    case SCAN,
         STOP_SCAN,
         CONNECT,
         DISCONNECT,
         DISCOVER_SERVICES,
         READ,
         WRITE,
         NOTIFICATIONS,
         CONNECTION_STATE,
         IS_CONNECTED,
         IS_ENABLED,
         DESTROY
}
